src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,136 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProject(_referencesMap, _lazyReverseReferencesMap, projectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProject(_transitiveReferencesMap, projectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProject(_reverseTransitiveReferencesMap, projectId);
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        internal ProjectDependencyGraph WithProjectReferenceRemoved(ProjectId projectId, ProjectId referencedProjectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+            Contract.ThrowIfFalse(_referencesMap[projectId].Contains(referencedProjectId));
+
+            // Removing a project reference doesn't change the set of projects
+            var projectIds = _projectIds;
+
+            // Incrementally update the graph
+            var referencesMap = _referencesMap.SetItem(projectId, _referencesMap[projectId].Remove(referencedProjectId));
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProjectReference(_lazyReverseReferencesMap, projectId, referencedProjectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProjectReference(_transitiveReferencesMap, projectId, referencedProjectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProjectReference(_reverseTransitiveReferencesMap, projectId, referencedProjectId);
+
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
```

Comment:
Do you want to keep the topological sort? Removing the reference won't invalidate the existing sort.

Reply:
I prefer not to change this for now. It could introduce non-determinism in the order even if both are technically correct, and it wasn't observed as a performance problem.